### ISSUE TITLE
Investigate possible hang during discovery

### ISSUE THREAD

**Issue by guruz** [owncloud/client#3122]
```
04-15 16:46:16:473 0x8bb290 OCC::SocketApi::sendMessage: SocketApi:  Sending message:  "STATUS:NEW:C:\Users\xxx\OwnCloud2\Music\LISTEN RETAG AND SORT & CHECK LEVELS\Mp3's\Music\E\Eric Johnson\Ah via musicom\zzz.mp3"
04-15 16:46:16:473 0x8bb290 OCC::SocketApi::command_RETRIEVE_FILE_STATUS: void OCC::SocketApi::command_RETRIEVE_FILE_STATUS(const QString&, QLocalSocket*) "C:\Users\xxx\OwnCloud2\Music\LISTEN RETAG AND SORT & CHECK LEVELS\Mp3's\Music\Incoming\Mary Mary\Incredible\yyy.mp3"
04-15 16:46:16:474 0x9bbcdb0 _csync_detect_update: Database entry found, compare: 1008103388 <-> 1008103388, etag: 8f91279ed7c5ba7213794d966585605e <-> 8f91279ed7c5ba7213794d966585605e, inode: 445127 <-> 445127, size: 4192634 <-> 4192634, perms:  <-> RDNVW
04-15 16:46:16:474 0x9bbcdb0 _csync_detect_update: file: Music/LISTEN RETAG AND SORT & CHECK LEVELS/Mp3's/Music/C/Collective Soul/xxx.mp3, instruction: INSTRUCTION_NONE <<=
... some more lines of csync_update in thread 0x9bbcdb0
```
Note how the thread 0x8bb290 does not continue with sending the result for RETRIEVE_FILE_STATUS
See also discussion on IRC.

The user reports that if he deinstalls the Explorer extension it works fine.
(Note that this not necessarily means this is OS-specific)

**Comment by eagles051387** [comment]
Suggest leaving the explorer extension disabled by default.

**Comment by ogoffart** [comment]
`SyncEngine::estimateState`  is O(n) in the number of files.  I believe it could be made O(log n) since the vector is sorted.  I don't know if this is the bottleneck.

**Comment by eagles051387** [comment]
From what I have seen it seems to be an issue with windows integration the
hang and crash. I have downoaded the 1.8.1 beta of the client and installed
it without explorer integration that eliminates the issue all together.
could there be a race condition maybe between discovery and windows
explorer integration?

On Fri, Apr 17, 2015 at 12:47 PM, Olivier Goffart <notifications@github.com>
wrote:

> SyncEngine::estimateState is O(n) in the number of files. I believe it
> could be made O(log n) since the vector is sorted. I don't know if this is
> the bottleneck.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/owncloud/client/issues/3122#issuecomment-93963310>.
>



-- 
Jonathan Aquilina

**Comment by ogoffart** [comment]
> could there be a race condition maybe between discovery and windows
explorer integration?

It's not a race condition.   The explorer integration asks for status icon to mirall, and this is really slow when there is a sync in progress with many file because we need to look up the state in a big vector.
This need to be optimized.

**Comment by dragotin** [comment]
ok. Moved to 1.9 Milestone.

**Comment by eagles051387** [comment]
Oliver as well besides optimized I think there must be memory leaks as well
as there is a massive amount of memory usage difference.

with explorer integration we are talking about 888mb of ram to 1.1gb before
this issue occurs

without - 444 - 555mb of ram usage.

On Tue, Apr 21, 2015 at 12:09 PM, Klaas Freitag <notifications@github.com>
wrote:

> ok. Moved to 1.9 Milestone.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/owncloud/client/issues/3122#issuecomment-94729546>.
>



-- 
Jonathan Aquilina

**Comment by jturcotte** [comment]
I pushed a few fixes that should avoid the torrent of RETRIEVE_FILE_STATUS. This should fix any noticeable slowdown during syncing, bringing the total from O(n2) to O(n) where n is the number of file synced.

It would be possible to do the binary search trick to bring this further down to O(log n), but it's also quite risky that the search would break at some point if the structure order isn't sorted for some reason, etc. So let's first see if it's still necessary.

**Comment by guruz** [comment]
OK let's close this and see if @eagles051387 gets problems when using 1.8.2

**Comment by eagles051387** [comment]
Would these fixes solve the issue with windows integration as that is where the issue is technically. I had tested it without windows integration and it works flawlessly

**Comment by guruz** [comment]
@eagles051387 Yes, do you want to try again with the integration?
http://download.owncloud.com/desktop/daily/ownCloud-1.8.1.5052-nightly20150507-setup.exe

**Comment by guruz** [comment]
We're assuming this works now...

**Comment by eagles051387** [comment]
Reopening this as I am testing the latest stable client on windows 10 tech
preview and i have loads of data to sync. Is there a newer version of the
client which has the windows explorer integration issues solved? as I am
seeing severe lockups in the form of the app crashing and the entire os
becoming unresponsive like loading file explorer. as soon as I killed the
OC client process they popped up. It seems to me there is a deadlock
somewhere.

On Wed, Jun 3, 2015 at 4:28 PM, Markus Goetz <notifications@github.com>
wrote:

> Closed #3122 <https://github.com/owncloud/client/issues/3122>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/owncloud/client/issues/3122#event-321340470>.
>



-- 
Jonathan Aquilina
